onnxruntime/core/framework/graph_partitioner.cc

```diff
@@ -230,4 +246,140 @@ Status GraphPartitioner::Partition(Graph& graph, bool export_dll, FuncManager& f
+  std::unordered_set<NodeIndex> visited;
+
+  std::unordered_set<const NodeArg*> cpu_output_args;
+  std::unordered_set<NodeIndex> provider_nodes;
+
+  for (auto& capability : capabilities) {
+    const IndexedSubGraph* indexed_sub_graph = capability->sub_graph.get();
+    if (indexed_sub_graph->GetMetaDef() != nullptr) {
+      continue;
+    }
+    // The <provider> can run a single node in the <graph> if not using meta-defs.
+    ORT_ENFORCE(1 == indexed_sub_graph->nodes.size());
+
+    const NodeIndex node_id = indexed_sub_graph->nodes[0];
+    provider_nodes.insert(node_id);
+    const Node* node = graph.GetNode(node_id);
+
+    const KernelCreateInfo* kernel_info;
+    Status st = provider->GetKernelRegistry()->TryFindKernel(*node, provider->Type(), &kernel_info);
+    ORT_ENFORCE(st.IsOK());
+
+    // first, find all the direct consumer of cpu tensors.
+    ORT_THROW_IF_ERROR(node->ForEachWithIndex(
+        node->OutputDefs(),
+        [&](const NodeArg& node_arg, size_t out_index) {
+          if (kernel_info->kernel_def->IsOutputOnCpu(out_index)) {
+            cpu_output_args.insert(&node_arg);
+            auto consumer_nodes = graph.GetConsumerNodes(node_arg.Name());
+            for (auto& consumer_node : consumer_nodes) {
+              candidates.push(consumer_node->Index());
+              LOGS_DEFAULT(INFO) << "Canditiate for fallback CPU execution: " << consumer_node->Name();
+            }
+          }
+          return Status::OK();
+        }));
+  }
+
+  const std::vector<const NodeArg*>& graph_inputs = graph.GetInputs();
+  std::unordered_set<NodeIndex> cpu_nodes;
+  // The algo below is trying to identity a subgraph that only depends on cpu tensors.
+  // Usually it is a subgraph that doing shape calculation based on a GPU tensor, then reshape it back.
+  // The detail:
+  // for each candidate, if one of its input is a cpu tensor and the cuda kernel doesn't mark it as cpu input,
+  // force the node to CPU to avoid memory cpu and add its output to the small cpu tensors.
+  while (!candidates.empty()) {
+    NodeIndex cur = candidates.top();
+    candidates.pop();
+    if (visited.count(cur) != 0)
+      continue;
+    visited.insert(cur);
+
+    if (provider_nodes.find(cur) == provider_nodes.end())
+      continue;
+
+    auto* node = graph.GetNode(cur);
+    // skip placing current node on CPU if no kernel is found
+    if (!KernelRegistryManager::HasImplementationOf(kernel_registry_mgr_, *node, kCpuExecutionProvider))
+      continue;
+
+    const KernelCreateInfo* kernel_info;
+    Status st = provider->GetKernelRegistry()->TryFindKernel(*node, provider->Type(), &kernel_info);
+
+    bool place_in_cpu = true;
+    for (size_t i = 0; i < node->InputDefs().size(); ++i) {
+      auto* input = node->InputDefs()[i];
+
+      // allow placing on CPU if it's a small initializer or graph input
+      if (IsSmallInitializer(graph, input) ||
```

Comment:
>IsSmallInitializer [](start = 10, length = 18)

To be safe, we should check all nodes that use the initializer, in case some of them are on GPU.

Reply:
done.